.circleci/test-scripts/baseline/main.py

```diff
@@ -0,0 +1,289 @@
+    with open(input_file_name, "r") as measure:
+        new_measurement = json.load(measure)
+
+        if data:
+            verify_new_data(data, new_measurement)
+
+        result = []
+        for _, values in group_data(data, "VmConfig", "CollectionMethod"):
+            ordered = sorted(values,
+                    key=itemgetter("Timestamp"), reverse=True)
+
+            # Each benchmark data contains two values, with and without
+            # collector, and the result array is a flattened version of it. The
+            # threshold is formulated in benchmarks, so double it.
+            if len(ordered) >= threshold * 2:
+                # Drop one oldest pair (no overhead, overhead) in every group
+                _, *no_overhead = filter(is_no_overhead, ordered)
+                _, *overhead = filter(is_overhead, ordered)
+            else:
+                no_overhead = [*filter(is_no_overhead, ordered)]
+                overhead = [*filter(is_overhead, ordered)]
+
+            result.extend(no_overhead + overhead)
+
+        result.extend(new_measurement)
+        return result
+
+
+def verify_new_data(baseline_data, new_data):
+    """
+    Enforce compatibility between baseline data and a new chunk of numbers. Two
+    invariants needs to be maintained:
+
+    * both baseline and new data have the same set of (VmConfig, CollectionMethod)
+
+    * New data provides equal number of with/without collector benchmark runs
+
+    The function expects baseline is not empty.
+
+    """
+
+    assert baseline_data, "Baseline data must be not empty"
+
+    baseline_grouped = process(baseline_data)
+    new_grouped = process(new_data)
+
+    for (baseline, new) in zip_longest(baseline_grouped, new_grouped):
+        bgroup, bvalues = baseline
+        ngroup, nvalues = new
+
+        if bgroup != ngroup:
+            raise Exception(f"Benchmark metrics do not match:"
+                             " {bgroup}, {ngroup} ")
+
+        counter = Counter()
+        for v in nvalues:
+            counter.update(v.keys())
+
+        no_overhead_count = counter.get("baseline_benchmark", 0)
+        overhead_count = counter.get("collector_benchmark", 0)
+
+        if (no_overhead_count != overhead_count):
+            raise Exception(f"Number of with/without overhead do not match:"
+                             " {no_overhead_count}, {overhead_count} ")
+
+
+def process(content):
+    """
+    Transform benchmark data into the format CI scripts work with.
+    """
+
+    data = [
+        record for record in content
+        if record.get("Metrics", {}).get("hackbench_avg_time")
+    ]
+
+    processed = [
+        {
+            "kernel": record.get("VmConfig"),
+            "collection_method": record.get("CollectionMethod"),
+            "timestamp": record.get("Timestamp"),
+            record["TestName"]: record.get("Metrics").get("hackbench_avg_time")
+        }
+        for record in data
+    ]
```

Comment:
Couldn't these two list comprehensions be combined if we moved the if condition of the first one to the last one?